_posts/2022-04-20-solving-issues-with-extensions.adoc

```diff
@@ -0,0 +1,247 @@
+except when you start using Mockito due to https://github.com/raphw/byte-buddy/issues/1162[a ByteBuddy issue]:
+ByteBuddy can get confused if there are several methods with the same signature but different return types.
+
+[NOTE]
+====
+ByteBuddy is an amazing library and do not see this post as a critic of it.
+This is an extreme corner case that doesn't happen with standard bytecode.
+====
+
+This issue was causing our tests to be unreliable because sometimes ByteBuddy was choosing the wrong method to apply Mockito magic.
+
+== How can we work around this?
+
+In the Quarkus GitHub App case, we don't really care about binary compatibility:
+when upgrading to a new version of the GitHub API,
+users will rebuild their application.
+
+So given these bridge methods are problematic,
+one solution would be to get rid of them.
+
+Obviously, we could fork the GitHub API and avoid generating the bridge methods.
+
+But forking and maintaining a fork forever is definitely not something we should consider if we can avoid it.
+Especially since we want to continue benefiting from all the future improvements of the GitHub API.
+
+So could we somehow keep the library standard but have Quarkus adjust the bytecode when building the application?
+
+If you are in a rush, the short answer is yes.
+Now let's go for the (not so) long answer.
+
+=== Let's identify the methods
+
+In Quarkus, we can index the annotations with Jandex so, in a perfect world,
+we would index the GitHub API jar with Jandex (which we already do for other purposes)
+and interrogate Jandex to get all the methods annotated with `@WithBridgeMethods`:
+
+[source,java,role=nowrap]
+----
+Collection<AnnotationInstance> withBridgeMethodsAnnotations =
+    index.getAnnotations(DotName.createSimple(WithBridgeMethods.class.getName));
+----
+
+Unfortunately, `@WithBridgeMethods` has a `CLASS` retention policy
+- which makes perfect sense for its usage -
+and Jandex only considers annotations with a `RUNTIME` retention policy.
+
+So, in this very case, we cannot use Jandex.
+
+Unfortunately, we don't have many options here:
+we have to list the methods manually.
+
+For more flexibility, we introduced a `BuildItem`:
+
+[source,java,role=nowrap]
+----
+public final class GitHubApiClassWithBridgeMethodsBuildItem extends MultiBuildItem {
+
+    private final String className;
+    private final Set<String> methodNames;
+
+    GitHubApiClassWithBridgeMethodsBuildItem(String className, String... methodsWithBridges) {
+        this.className = className;
+        this.methodNames = new HashSet<>(Arrays.asList(methodsWithBridges));
+    }
+
+    public String getClassName() {
+        return className;
+    }
+
+    public Set<String> getMethodsWithBridges() {
+        return methodNames;
+    }
+}
+----
+
+And we will produce a `GitHubApiClassWithBridgeMethodsBuildItem` for each class:
+
+[source,java,role=nowrap]
+----
+// ...
+
+classesWithBridgeMethods.produce(new GitHubApiClassWithBridgeMethodsBuildItem(
```

Comment:
Maybe it's not 100% clear to me, but because the WithBridgeMethid is a class retention policy you can't perform reflection from the sources and need to list them manually (performing changes when a new Hub4j version is out)

Reply:
Yeah I didn't want to go into reflection as it would have been a huge hack anyway. And you would have to list the class anyway if you don't want to go through all of them.